src/io/port.rs

```diff
@@ -0,0 +1,94 @@
+//! Solaris/Illumos event ports.
+
+use crate::backend::{c, io::syscalls};
+use crate::fd::{AsFd, AsRawFd, OwnedFd};
+use crate::io;
+
+use super::PollFlags;
+
+use core::time::Duration;
+
+/// The structure representing a port event.
+#[repr(transparent)]
+pub struct Event(pub(crate) c::port_event);
+
+impl Event {
+    /// Get the events associated with this event.
+    pub fn events(&self) -> i32 {
+        self.0.portev_events
+    }
+
+    /// Get the event source associated with this event.
+    pub fn object(&self) -> usize {
+        self.0.portev_object
+    }
+
+    /// Get the userdata associated with this event.
+    pub fn userdata(&self) -> *mut () {
+        self.0.portev_user.cast()
+    }
+}
+
+/// `port_create()`- Creates a new port.
+pub fn port_create() -> io::Result<OwnedFd> {
+    syscalls::port_create()
+}
+
+/// `port_associate()`- Associates a file descriptor with a port.
+pub fn port_associate_fd(
+    port: &impl AsFd,
+    object: &impl AsFd,
+    events: PollFlags,
+    userdata: *mut (),
+) -> io::Result<()> {
+    unsafe {
+        syscalls::port_associate(
+            port.as_fd(),
+            c::PORT_SOURCE_FD,
+            object.as_fd().as_raw_fd() as _,
+            events.bits() as _,
+            userdata.cast(),
+        )
+    }
+}
+
+/// `port_dissociate()`- Dissociates a file descriptor from a port.
+pub fn port_dissociate_fd(port: &impl AsFd, object: &impl AsFd) -> io::Result<()> {
+    unsafe {
+        syscalls::port_dissociate(
+            port.as_fd(),
+            c::PORT_SOURCE_FD,
+            object.as_fd().as_raw_fd() as _,
+        )
+    }
+}
+
+/// `port_get()`- Gets an event from a port.
+pub fn port_get(port: &impl AsFd, timeout: Option<Duration>) -> io::Result<Option<Event>> {
+    let mut timeout = timeout.map(|timeout| c::timespec {
+        tv_sec: timeout.as_secs() as _,
+        tv_nsec: timeout.subsec_nanos() as _,
+    });
+
+    syscalls::port_get(port.as_fd(), timeout.as_mut())
+}
+
+/// `port_getn()`- Gets multiple events from a port.
+pub fn port_getn(
+    port: &impl AsFd,
+    events: &mut Vec<Event>,
+    min_events: usize,
+    timeout: Option<Duration>,
+) -> io::Result<()> {
+    let mut timeout = timeout.map(|timeout| c::timespec {
+        tv_sec: timeout.as_secs() as _,
```

Comment:
Can the seconds conversion overflow? If so, please use `try_into()` and fail as needed.